### DISCUSSION TITLE
Equivalent of TikZ's externalize functionality

### DISCUSSION THREAD

**Discussion by SolidTux** [cetz-package/cetz#525] [Ideas]
Currently, I have a project that when compiled in once causes Typst's memory usage to explode. I solved it by putting it into a separate file and then importing the SVG. I think this could be solved with the help of an external tool. But there are many unclear things to me before even trying to start some prototype:

* Should it respect the currently active `show` and `set` rules? (e.g. should a previous `set text(...)` change the font?)
* Is that even possible in any feasible way?
* How should the building be handled? One option would be a Makefile, but that doesn't feel well integrated into the Typst build process.
* How should relative sizes (both `%` and `em`) be handled?

As a first step, one could preprocess the Typst file and just stupidly put the `canvas` in a separate file, prefix it with the imports and a `set page(width: auto, height: auto, margin: 0pt)` and replace the `canvas` with `image("file.svg")`.

That's not a great solution, though, as it's definitely expected that the document will look the same with and without externalize. Maybe with the upcoming context changes, it would be possible to make one dummy run that extracts all properties like `text.font`, the container size, styles, states, ..., saves them in a JSON, YAML or whatever and then the separate file could use this information.

**Comment by fenjalien** [comment]
Could you share the project or at least a minimum working example? I know CeTZ isn't exactly optimised but this sounds like more of a Typst issue/feature request than just CeTZ...

**Reply by SolidTux** [reply]
I'll prepare a smaller example, where it still appears as it currently is a full conference poster (partially my own dumb logic to make blocks might be a problem that adds unnecessary complexity).

**Reply by SolidTux** [reply]
I couldn't reproduce the extreme memory usage, but [here is the poster](https://gitlab.com/solidtux-phd/dpg24-poster). Probably I had more points for the plots or something like that, I'll experiment with that and report back.

**Comment by johannes-wolf** [comment]
I quite like the idea. Cetz could provide something like `cezt.external(<path or source-code>)`, an external tool queries those call using Typst metadata and `query` command, compiles them to svg and places the results in a special dir. A hash of the path/filename/code could be used as an identifier/name for the svg. The metadata can also include a json of the set show-rules.

**Reply by SolidTux** [reply]
> Cetz could provide something like cezt.external(<path or source-code>),

True, probably as a first step, manually creating separate files and inserting the externalize command might be the easiest solution.

> A hash of the path/filename/code could be used as an identifier/name for the svg.

That might be not that ideal as by just changing the code one would end up with a lot of temporary files. In general, it would be nice to have an option to specify the filename manually.

**Reply by johannes-wolf** [reply]
I wrote a poc implementation in Python which kinda works. It is quite hacky though, as Typst provides no way to check if a file exists. The workaround is, that the `external` tool sets a state variable to skip `image` calls when querying the metadata. You can check it out here: https://github.com/johannes-wolf/cetz/tree/cetz-external-poc -- The `external` command assumes that the current "@preview/cetz:0.2.1" has the functionality, so you'd have to run `just install @preview` to install the patched cetz version locally.

**Reply by fenjalien** [reply]
I've had two ideas about alternate work arounds:
1. You use a template function that places all the set rules and you use it with a global show rule: `#show: template`. You then import the template and use the show rule at the top of each file.
2. If you have Typst 0.11 (the release candidate would work) you could use the new `sys.inputs` dictionary to have a "mode" input. In one mode you generate all the required diagrams on separate pages, they'll get split into separate svg files. Then in another mode you generate the document normally and insert the diagram svgs with `image`.

**Reply by SolidTux** [reply]
Typst had quite some updates since then. Is there something that would make this more feasible?
